plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -2384,6 +2419,74 @@ public void finishStatisticsCollection(ConnectorSession session, ConnectorTableH
+                        || addFileEntry.getStats().get().getMaxValues().isEmpty()
+                        || addFileEntry.getStats().get().getMinValues().isEmpty()
+                        || addFileEntry.getStats().get().getNullCount().isEmpty())
+                // Statistics returns whole path to file, so we need to create corresponding map key for AddFileEntry.
+                .map(addFileEntry -> Map.entry(appendPath(tableHandle.getLocation(), addFileEntry.getPath()), addFileEntry))
+                .collect(Collectors.toUnmodifiableMap(Entry::getKey, Entry::getValue));
+
+        if (addFileEntriesWithNoStats.isEmpty()) {
+            return;
+        }
+
+        Map<String, DeltaLakeColumnHandle> columnsHandles = getColumns(tableHandle.getMetadataEntry()).stream()
+                .filter(column -> column.getColumnType() == REGULAR)
+                .map(column -> Map.entry(column.getName(), column))
+                .collect(Collectors.toUnmodifiableMap(Entry::getKey, Entry::getValue));
+
+        List<AddFileEntry> updatedAddFileEntries = computedStatistics.stream()
+                .map(statistics -> {
+                    // Grouping by `PATH_COLUMN_NAME`.
+                    String filePathFromStatistics = VARCHAR.getSlice(statistics.getGroupingValues().get(0), 0).toStringUtf8();
+                    if (addFileEntriesWithNoStats.containsKey(filePathFromStatistics)) {
+                        return Optional.of(prepareUpdatedAddFileEntry(statistics, addFileEntriesWithNoStats.get(filePathFromStatistics), columnsHandles));
+                    }
+                    return Optional.<AddFileEntry>empty();
+                })
+                .flatMap(Optional::stream)
+                .toList();
+
+        if (!updatedAddFileEntries.isEmpty()) {
+            try {
+                long createdTime = Instant.now().toEpochMilli();
+                long readVersion = tableHandle.getReadVersion();
+                long commitVersion = readVersion + 1;
+                TransactionLogWriter transactionLogWriter = transactionLogWriterFactory.newWriter(session, tableHandle.getLocation());
+                transactionLogWriter.appendCommitInfoEntry(getCommitInfoEntry(session, commitVersion, createdTime, OPTIMIZE_OPERATION, readVersion));
+                updatedAddFileEntries.forEach(transactionLogWriter::appendAddFileEntry);
+                transactionLogWriter.flush();
+            }
+            catch (Throwable e) {
+                throw new TrinoException(DELTA_LAKE_BAD_WRITE, "Unable to access file system for: " + tableHandle.getLocation(), e);
+            }
+        }
+    }
+
+    private AddFileEntry prepareUpdatedAddFileEntry(ComputedStatistics stats, AddFileEntry addFileEntry, Map<String, DeltaLakeColumnHandle> columnsHandles)
+    {
+        DeltaLakeJsonFileStatistics deltaLakeJsonFileStatistics = DeltaLakeComputedStatistics.toDeltaLakeJsonFileStatistics(stats, columnsHandles);
```

Comment:
We could have a `withStats(Optional<String> stats)` in `AddFileEntry` to do the copying of the content of the existing `AddFileEntry` with stats. 

If we inline within `generateMissingFileStatistics(..)` method the calcuation of `DeltaLakeJsonFileStatistics deltaLakeJsonFileStatistics`, then the method `prepareUpdatedAddFileEntry` is not needed anymore.